functions/transformations/assert_equals.go

```diff
@@ -0,0 +1,239 @@
+	}
+
+	return &AssertEqualsOpSpec{Name: name}, nil
+}
+
+func newAssertEqualsOp() flux.OperationSpec {
+	return new(AssertEqualsOpSpec)
+}
+
+type AssertEqualsProcedureSpec struct {
+	plan.DefaultCost
+	Name string
+}
+
+func (s *AssertEqualsProcedureSpec) Kind() plan.ProcedureKind {
+	return AssertEqualsKind
+}
+
+func (s *AssertEqualsProcedureSpec) Copy() plan.ProcedureSpec {
+	return &AssertEqualsProcedureSpec{}
+}
+
+func newAssertEqualsProcedure(qs flux.OperationSpec, pa plan.Administration) (plan.ProcedureSpec, error) {
+	spec, ok := qs.(*AssertEqualsOpSpec)
+	if !ok {
+		return nil, fmt.Errorf("invalid spec type %T", qs)
+	}
+	return &AssertEqualsProcedureSpec{Name: spec.Name}, nil
+}
+
+type AssertEqualsTransformation struct {
+	mu sync.Mutex
+
+	gotParent  *assertEqualsParentState
+	wantParent *assertEqualsParentState
+
+	d     execute.Dataset
+	cache execute.TableBuilderCache
+
+	name string
+}
+
+type assertEqualsParentState struct {
+	id         execute.DatasetID
+	mark       execute.Time
+	processing execute.Time
+	finished   bool
+}
+
+func createAssertEqualsTransformation(id execute.DatasetID, mode execute.AccumulationMode, spec plan.ProcedureSpec, a execute.Administration) (execute.Transformation, execute.Dataset, error) {
+	if len(a.Parents()) != 2 {
+		return nil, nil, errors.New("assertEquals should have exactly 2 parents")
+	}
+
+	cache := execute.NewTableBuilderCache(a.Allocator())
+	dataset := execute.NewDataset(id, mode, cache)
+	pspec, ok := spec.(*AssertEqualsProcedureSpec)
+	if !ok {
+		return nil, nil, fmt.Errorf("invalid spec type %T", pspec)
+	}
+
+	transform := NewAssertEqualsTransformation(dataset, cache, pspec, a.Parents()[0], a.Parents()[1])
+
+	return transform, dataset, nil
+}
+
+func NewAssertEqualsTransformation(d execute.Dataset, cache execute.TableBuilderCache, spec *AssertEqualsProcedureSpec, gotID, wantID execute.DatasetID) *AssertEqualsTransformation {
+	return &AssertEqualsTransformation{
+		gotParent:  &assertEqualsParentState{id: gotID},
+		wantParent: &assertEqualsParentState{id: wantID},
+		d:          d,
+		cache:      cache,
+		name:       spec.Name,
+	}
+}
+
+func (t *AssertEqualsTransformation) RetractTable(id execute.DatasetID, key flux.GroupKey) error {
+	panic("not implemented")
+}
+
+func (t *AssertEqualsTransformation) Process(id execute.DatasetID, tbl flux.Table) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	var colMap = make([]int, 0, len(tbl.Cols()))
+	var err error
+	builder, created := t.cache.TableBuilder(tbl.Key())
+	if created {
+		colMap, err = execute.AddNewTableCols(tbl, builder, colMap)
+		if err != nil {
+			return err
+		}
+		if err := execute.AppendMappedTable(tbl, builder, colMap); err != nil {
+			return err
+		}
+	} else {
+		cacheTable, err := builder.Table()
+		if err != nil {
+			return err
+		}
+		if ok, err := execute.TablesEqual(cacheTable, tbl); !ok {
+			return fmt.Errorf("test %s: tables not equal", t.name)
+		} else if err != nil {
+			return err
```

Comment:
Should the error check happen first? The way it is written this line is unreachable, since any error returned from execute.TablesEqual is also !ok,